File: apps/api/app/api/routes/extraction.py
"""Document extraction and candidate-review endpoints."""

from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database.enums import CandidateStatus
from app.database.session import get_db
from app.extraction.service import ExtractionService, get_extraction_service
from app.schemas.candidate import (
    CandidateCorrectionRequest,
    CandidateDetail,
    CandidateReviewResult,
    ExtractionSummary,
)

router = APIRouter(tags=["extraction"])


@router.post("/documents/{document_id}/extract", response_model=ExtractionSummary)
def extract_document(
    document_id: str,
    db: Annotated[Session, Depends(get_db)],
    service: Annotated[ExtractionService, Depends(get_extraction_service)],
) -> ExtractionSummary:
    return service.extract(db, document_id)


@router.get("/documents/{document_id}/candidates", response_model=list[CandidateDetail])
def list_candidates(
    document_id: str,
    db: Annotated[Session, Depends(get_db)],
    service: Annotated[ExtractionService, Depends(get_extraction_service)],
    candidate_status: Annotated[CandidateStatus | None, Query(alias="status")] = None,
) -> list[CandidateDetail]:
    return service.list_candidates(db, document_id, candidate_status)


@router.post("/candidates/{candidate_id}/accept", response_model=CandidateReviewResult)
def accept_candidate(
    candidate_id: str,
    db: Annotated[Session, Depends(get_db)],
    service: Annotated[ExtractionService, Depends(get_extraction_service)],
) -> CandidateReviewResult:
    return service.accept(db, candidate_id)


@router.patch("/candidates/{candidate_id}/correct", response_model=CandidateReviewResult)
def correct_candidate(
    candidate_id: str,
    correction: CandidateCorrectionRequest,
    db: Annotated[Session, Depends(get_db)],
    service: Annotated[ExtractionService, Depends(get_extraction_service)],
) -> CandidateReviewResult:
    return service.correct(db, candidate_id, correction)


@router.post("/candidates/{candidate_id}/reject", response_model=CandidateReviewResult)
def reject_candidate(
    candidate_id: str,
    db: Annotated[Session, Depends(get_db)],
    service: Annotated[ExtractionService, Depends(get_extraction_service)],
) -> CandidateReviewResult:
    return service.reject(db, candidate_id)

File: apps/web/lib/api/extraction.ts
import { apiGet, apiMutation, hasString, isArrayOf, isRecord } from "./client";
import type {
  CandidateReviewResult,
  CandidateStatus,
  ExtractedCandidate,
  ExtractionSummary,
} from "@/lib/types/api";

function isCandidate(value: unknown): value is ExtractedCandidate {
  return (
    isRecord(value) &&
    hasString(value, "id") &&
    hasString(value, "document_id") &&
    hasString(value, "candidate_type") &&
    isRecord(value.structured_data) &&
    typeof value.confidence === "number" &&
    hasString(value, "status") &&
    hasString(value, "evidence_text") &&
    (value.page_number === null || typeof value.page_number === "number") &&
    hasString(value, "created_at") &&
    hasString(value, "updated_at")
  );
}

function isExtractionSummary(value: unknown): value is ExtractionSummary {
  return (
    isRecord(value) &&
    hasString(value, "document_id") &&
    value.status === "completed" &&
    typeof value.candidate_count === "number" &&
    isRecord(value.by_type) &&
    typeof value.reused_existing === "boolean"
  );
}

function isReviewResult(value: unknown): value is CandidateReviewResult {
  return isRecord(value) && isCandidate(value.candidate) &&
    (value.trusted_record === null || isRecord(value.trusted_record));
}

export function extractDocument(documentId: string): Promise<ExtractionSummary> {
  return apiMutation(
    `/api/v1/documents/${encodeURIComponent(documentId)}/extract`,
    "POST",
    isExtractionSummary,
  );
}

export function getCandidates(
  documentId: string,
  status?: CandidateStatus,
): Promise<ExtractedCandidate[]> {
  const query = status ? `?status=${encodeURIComponent(status)}` : "";
  return apiGet(
    `/api/v1/documents/${encodeURIComponent(documentId)}/candidates${query}`,
    (value): value is ExtractedCandidate[] => isArrayOf(value, isCandidate),
  );
}

export function acceptCandidate(candidateId: string): Promise<CandidateReviewResult> {
  return apiMutation(
    `/api/v1/candidates/${encodeURIComponent(candidateId)}/accept`,
    "POST",
    isReviewResult,
  );
}

export function correctCandidate(
  candidateId: string,
  structuredData: Record<string, unknown>,
): Promise<CandidateReviewResult> {
  return apiMutation(
    `/api/v1/candidates/${encodeURIComponent(candidateId)}/correct`,
    "PATCH",
    isReviewResult,
    { structured_data: structuredData },
  );
}

export function rejectCandidate(candidateId: string): Promise<CandidateReviewResult> {
  return apiMutation(
    `/api/v1/candidates/${encodeURIComponent(candidateId)}/reject`,
    "POST",
    isReviewResult,
  );
}
